### ISSUE TITLE
Configuration for LSUN Church 256 experiments

### ISSUE THREAD

**Issue by Antollo** [SHI-Labs/StyleNAT#18]
Hi, thank you for this great work!

I'm trying to reproduce the LSUN Church 256 results from the paper. I noticed that the repo only includes `conf/runs/ffhq_256.yaml`. From reading the paper (Section 4.2 and Appendix F.2.2), I understand the key changes from FFHQ were:
- Increasing `min_heads` from 4 to 8 (`runs.generator.min_heads=8`)
- Using pyramid (progressive) dilations. Would the correct configuration be `runs.generator.dilations='[[1],[1],[1,2],[1,2,4],[1,2,4,8],[1,2,4,8,16],[1,2,4,8,16,32]]`?
- Does the kernel size need to be reduced to 3, or does it remain the same?

I just wanted to confirm if my understanding of these details is correct. Additionally, were any other training parameters (learning rate, batch size, etc.) changed from the FFHQ setup for Church experiments?

Thank you in advance for your help!

**Comment by stevenwalton** [comment]
Hey Antollo, I don't seem to have the exact run config anymore but I believe it should look something like this

```yaml
size: 256

generator:
    n_mlp: 8
    block_type: nat
    style_dim: 512
    lr: 0.00004
    channel_multiplier: 2
    mlp_ratio: 4
    use_checkpoint: False
    lr_mlp: 0.01
    enable_full_resolution: 8
    min_heads: 8
    qkv_bias: True
    qk_scale: null
    proj_drop: 0.
    attn_drop: 0.
    kernels: [[3],[7],[7],[7],[7],[7],[7],[7],[7]]
    dilations: [[1],[1],[1,2],[1,2,4],[1,2,4,8],[1,2,4,8,16],[1,2,4,8,16,32]
    reg_every: null
    params: 0

discriminator:
    lr: 0.0002
    channel_multiplier: 2
    blur_kernel: [1, 3, 3, 1]
    sn: True
    ssd: False
    reg_every: 16
    params: 0

training:
    iter: 1000000
    batch: 8
    use_flip: True
    ttur: True
    r1: 20
    bcr: True
    bcr_fake_lambda: 10
    bcr_real_lambda: 10
    beta1: 0.0
    beta2: 0.99
    start_dim: 512
    workers: 8
    lr_decay: True
    lr_decay_start_steps: 500000
    gan_weight: 1
```

Your understanding is correct. Note that with this config it'll bias 2 heads (we have 6 unique heads but 8 total). The only other things changed were the r1 value (I believe `15` was also a good value) and the `lr_decay_start_steps`. I will say that LSUN was a bit trickier of a run and it was frequently unstable. Something I never got the chance to really test was sweeping over the kernel sizes (I was pretty compute bound), so all those stayed the same. 

FWIW, I do think you could beat my result with a different configuration. If you want to go for this I'd try 3 things.
1) Shift to progressive training. This should help speed things up too
2) Use larger kernels at smaller resolutions or shift more towards larger effective kernels at the lower resolutions.
3) bigger effective kernels at lower resolutions and reducing number of small/medium heads at higher resolutions

Look at the attention maps in the appendix and you'll see the different heads do some different things. Like at 256 the undilated/low-dilated heads produce scan lines but the fully dilated heads capture more edges. I do think you will still want an undilated head to help capture detail but I think the errors are propagating up from lower resolution, so it is not able to effectively do its job. 

It may be a good idea to output the attention maps every so often since they can help you keep track of when things are becoming unstable. The maps were something created later in the project so I never got to take full advantage of them while training, but they will help you watch for stability issues. 

Let me know if you have any other questions or need any other help. I'd love to see how the reproduction goes

**Comment by Antollo** [comment]
Thanks for the fast response! The `r1` value must have been the detail I was missing (I got FID of ~7 in my early runs and the r1_val was rapidly growing). I really appreciate you sharing the full config and the extra tips on kernel sizes and attention maps. I'll be sure to update you on how the reproduction goes.

**Comment by stevenwalton** [comment]
Great! I'll close for now, but feel free to open back up or create any new issues.
